crates/miden-objects/src/block/proposed_block.rs

```diff
@@ -0,0 +1,445 @@
+                .insert(update.initial_state_commitment(), (update, batch.id()))
+            {
+                return Err(ProposedBlockError::ConflictingBatchesUpdateSameAccount {
+                    account_id,
+                    initial_state_commitment: conflicting_update.initial_state_commitment(),
+                    first_batch_id: conflicting_batch_id,
+                    second_batch_id: batch.id(),
+                });
+            }
+        }
+    }
+
+    // Build account witnesses.
+    let mut account_witnesses = Vec::with_capacity(updated_accounts.len());
+    let mut block_updates = Vec::with_capacity(updated_accounts.len());
+
+    for (account_id, mut updates) in updated_accounts {
+        let (initial_state_commitment, proof) = block_inputs
+            .accounts_mut()
+            .remove(&account_id)
+            .map(|witness| witness.into_parts())
+            .ok_or(ProposedBlockError::MissingAccountWitness(account_id))?;
+
+        let mut details: Option<AccountUpdateDetails> = None;
+
+        // Chronologically chain updates for this account together using the state hashes to
+        // link them.
+        let mut transactions = Vec::new();
+        let mut current_commitment = initial_state_commitment;
+        while !updates.is_empty() {
+            let (update, _) = updates.remove(&current_commitment).ok_or_else(|| {
+                ProposedBlockError::InconsistentAccountStateTransition(
+                    account_id,
+                    current_commitment,
+                    updates.keys().copied().collect(),
+                )
+            })?;
+
+            current_commitment = update.final_state_commitment();
+            let (update_transactions, update_details) = update.into_parts();
+            transactions.extend(update_transactions);
+
+            details = Some(match details {
+                None => update_details,
+                Some(details) => details.merge(update_details).map_err(|source| {
+                    ProposedBlockError::AccountUpdateError { account_id, source }
+                })?,
+            });
+        }
+
+        account_witnesses.push((
+            account_id,
+            AccountUpdateWitness {
+                initial_state_commitment,
+                final_state_commitment: current_commitment,
+                initial_state_proof: proof,
+                transactions: core::mem::take(&mut transactions),
+            },
+        ));
+
+        block_updates.push(BlockAccountUpdate::new(
+            account_id,
+            current_commitment,
+            details.expect("Must be some by now"),
+            transactions,
+        ));
+    }
+
+    Ok((account_witnesses, block_updates))
+}
+
+/// TODO
+#[derive(Debug, PartialEq, Eq)]
+pub struct AccountUpdateWitness {
+    initial_state_commitment: Digest,
+    final_state_commitment: Digest,
+    initial_state_proof: MerklePath,
+    transactions: Vec<TransactionId>,
+}
+
+impl AccountUpdateWitness {
```

Comment:
Would it make sense to add something like `merge()` method to this so that we could build `AccountUpdateWitness` incrementally? This could look something like:
```rust
pub fn merge(&mut self, update: &BatchAccountUpdate) -> Result<(), SomeError> {
    ...
}
```

Reply:
It's a good suggestion, but the difficulty is that `AccountUpdateWitness` can't be initialized properly, because it would have to contain `AccountUpdateDetails` (which it currently doesn't but should). There is no `AccountUpdateDetails::default()` impl and so we'd have to deal with `Option` here or we'd have to make a `AccountUpdateWitnessBuilder` or something, which is overall less nice than having the logic inlined in `aggregate_account_updates` I think.